### DISCUSSION TITLE
OT Dataset vs channel/panid/network key

### DISCUSSION THREAD

**Discussion by farhangj** [openthread/openthread#7820] [Q&A] [answered]
Hello OT community,

I'm working on an application based on OT and have some questions regarding OT operational dataset:
In the application in one network, I have multiple devices, some FTD and some MTD. "commissioning" is done via a custom out-of-band method between devices (not over Thread) where I share the OT Op dataset between the devices. 
I could have multiple networks (i.e. with different OT datasets) that coexist in one area.

- First of all is there something logically wrong with this out-of-band commissioning method?
- Is changing "network key" enough to isolate networks?

To change the Op dataset, in the OT API there are dataset API such as otDatasetSetActive() and lower level: direct Link settings API such as otLinkSetPanId() or otThreadSetNetworkKey().

- I assume I should be using the dataset API rather than setting the lower level settings - correct?
- When/in what conditions does the Op dataset get saved in NV memory for an MTD or FTD. e.g. does MTD wait till it has a parent before committing/saving this dataset?
- If one changed the settings directly via the otLink or otThread API, or CLI (e.g. ot channel <n>), when would that get reflected into the operational dataset?
- if I create a new active dataset on a parent that already has children, does it convey the changes to the children before moving on to the new settings - or do I have to convey this using my custom OoB method?

**Comment by jwhui** [comment, marked as answer]
> First of all is there something logically wrong with this out-of-band commissioning method?

No, there is nothing wrong commissioning out-of-band. See [src/cli/README_DATASET.md](https://github.com/openthread/openthread/blob/main/src/cli/README_DATASET.md) for an example of out-of-band commissioning.

> Is changing "network key" enough to isolate networks?

For the most part yes. The only consideration is that devices on the same channel and PAN ID may use the same IEEE 802.15.4 Short Address, which can cause issues with IEEE 802.15.4 Acknowledgments. However, this can happen with any IEEE 802.15.4-based network, including Zigbee.

> I assume I should be using the dataset API rather than setting the lower level settings - correct?

Yes.

> When/in what conditions does the Op dataset get saved in NV memory for an MTD or FTD. e.g. does MTD wait till it has a parent before committing/saving this dataset?

Calling `otDatasetSetActive` should always save to non-volatile.

> If one changed the settings directly via the otLink or otThread API, or CLI (e.g. ot channel ), when would that get reflected into the operational dataset?

Using these APIs are mainly for RCP and/or testing modes. They will not update the operational dataset.

> if I create a new active dataset on a parent that already has children, does it convey the changes to the children before moving on to the new settings - or do I have to convey this using my custom OoB method?

If you want to propagate an operational dataset change that will break connectivity (e.g. changing network key), then set use the Pending Operational Dataset feature. The Pending Operational Dataset includes a Delay Timer value that indicates when the new values should take affect, allowing for the Pending Operational Dataset to disseminate throughout the Thread network. Note that you should advance the Active Timestamp when updating the dataset.
